src/ut/bloom_filter_test.cpp

```diff
@@ -0,0 +1,200 @@
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or (at your
+ * option) any later version, along with the "Special Exception" for use of
+ * the program along with SSL, set forth below. This program is distributed
+ * in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+ * A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details. You should have received a copy of the GNU General Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ *
+ * The author can be reached by email at clearwater@metaswitch.com or by
+ * post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+ *
+ * Special Exception
+ * Metaswitch Networks Ltd  grants you permission to copy, modify,
+ * propagate, and distribute a work formed by combining OpenSSL with The
+ * Software, or a work derivative of such a combination, even if such
+ * copying, modification, propagation, or distribution would otherwise
+ * violate the terms of the GPL. You must comply with the GPL in all
+ * respects for all of the code used other than OpenSSL.
+ * "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+ * Project and licensed under the OpenSSL Licenses, or a work based on such
+ * software and licensed under the OpenSSL Licenses.
+ * "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+ * under which the OpenSSL Project distributes the OpenSSL toolkit software,
+ * as those licenses appear in the file LICENSE-OPENSSL.
+ */
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include "bloom_filter.h"
+
+using ::testing::Return;
+using ::testing::_;
+
+TEST(BloomFilterTest, NewBloomFilterIsEmpty)
+{
+  BloomFilter bf(10000, 1);
+  EXPECT_FALSE(bf.check("Kermit"));
+  EXPECT_FALSE(bf.check("MissPiggy"));
+  EXPECT_FALSE(bf.check("Gonzo"));
+  EXPECT_FALSE(bf.check("Animal"));
+}
+
+TEST(BloomFilterTest, OneBitPerItem)
+{
+  BloomFilter bf(10000, 1);
+
+  bf.add("Kermit");
+  bf.add("MissPiggy");
+
+  EXPECT_TRUE(bf.check("Kermit"));
+  EXPECT_TRUE(bf.check("MissPiggy"));
+  EXPECT_FALSE(bf.check("Gonzo"));
+  EXPECT_FALSE(bf.check("Animal"));
+}
+
+TEST(BloomFilterTest, TwoBitsPerItem)
+{
+  BloomFilter bf(10000, 2);
+
+  bf.add("Kermit");
+  bf.add("MissPiggy");
+
+  EXPECT_TRUE(bf.check("Kermit"));
+  EXPECT_TRUE(bf.check("MissPiggy"));
+  EXPECT_FALSE(bf.check("Gonzo"));
+  EXPECT_FALSE(bf.check("Animal"));
+}
+
+TEST(BloomFilterTest, ManyBitsPerItem)
+{
+  BloomFilter bf(10000, 10);
+
+  bf.add("Kermit");
+  bf.add("MissPiggy");
+
+  EXPECT_TRUE(bf.check("Kermit"));
+  EXPECT_TRUE(bf.check("MissPiggy"));
+  EXPECT_FALSE(bf.check("Gonzo"));
+  EXPECT_FALSE(bf.check("Animal"));
+}
+
+TEST(BloomFilterTest, NumEntriesAndFalsePositiveRate)
+{
+  BloomFilter* bf = BloomFilter::for_num_entries_and_fp_prob(2, 0.0001);
```

Comment:
Well the false positive probability for this test should be 0.01% (that's the 2nd argument to the function). But I agree that any of these tests can spuriously fail, but I don't think there is much that can be done to prevent this - bloom filters are probabilistic after all! With a rate of 0.01% I reckon we'll have fewer than one spurious failure a year, which feels acceptable. I'll add a comment to the file though.